docs/json.adoc

```diff
@@ -0,0 +1,156 @@
+void singleCustomer(Customer customer) {
+    assertThat(Collections.singleton(tuple(customer.getName(), customer.getHeight())))
+            .containsAnyOf(
+                tuple("Luke", 172),
+                tuple("Yoda", 66)
+            );
+}
+----
+
+This parametrized test will generate the following tests:
+
+* [1] Customer{name='Luke', height=172}
+* [2] Customer{name='Yoda', height=66}
+
+It is also possible to extract only a single element from each argument object by using the `@Param` annotation.
+
+[source,java]
+----
+@ParametrizedTest
+@JsonFileSource(resources = "/customers.json")
+void singleCustomer(@Param("name") String customerName) {
+    assertThat(customername)
+            .isIn("Luke", "Yoda");
+}
+----
+
+This parametrized test will generate the following tests:
+
+* [1] Luke
+* [2] Yoda
+
+=== Multiple Argument Methods
+
+When the method has multiple arguments then each json object argument will be deconstructed to each of the method arguments.
+By default, the method argument name will be used for locating the element that needs to be taken from the json object.
+You can also use `@Param` to give the name of the element that needs to be extracted.
+
+[IMPORTANT]
+====
+If your test sources are not compiled using the `--parameters` flag then the names of the arguments will not be like they are written in the source code.
+In that the situation you need to use `@Param` instead.
+====
+
+Using the same `customers.json` and the following test
+
+[source,java]
+----
+@ParametrizedTest
+@JsonFileSource(resources = "/customers.json")
+void deconstructFromArray(@Param("name") String customerName, @Param("height") int height) {
+    assertThat(Collections.singleton(tuple(customerName, height)))
+					.containsAnyOf(
+							tuple("Luke", 172),
+							tuple("Yoda", 66)
+					);
+}
+----
+
+This parametrized test will generate the following tests:
+
+* [1] Luke, 172
+* [2] Yoda, 66
+
+=== Extracting nested array
+
+Sometimes we want to extract a nested array instead of the root element.
+For this purpose `JsonFileSource#data` can be used.
+
+.Customer with nested array
+[source,json]
+----
+{
+  "name": "Luke",
+  "height": 172,
+  "vehicles": [
+    {
+      "name": "Snowspeeder",
+      "length": 4.5
+    },
+    {
+      "name": "Imperial Speeder Bike",
+      "length": 3
+    }
+  ]
+}
+----
+
+Here we want to test the customer vehicles.
+The test for this will look like:
+
+[source,java]
+----
+@ParameterizedTest
+@JsonFileSource(resources = "customer-luke.json", data = "vehicles")
+void lukeVehicles(@Param("name") String name, @Param("length") double length) {
+    assertThat(Collections.singleton(tuple(name, length)))
+            .containsAnyOf(
+                tuple("Snowspeeder", 4.5),
+                tuple("Imperial Speeder Bike", 3d)
+            );
+}
+----
+
+This parametrized test will generate the following tests:
+
+* [1] Snowspeeder, 4.5
+* [2] Imperial Speeder Bike, 3
```

Comment:
Can you add a section on thread safety like the other extensions have? Will probably look [like this](https://junit-pioneer.org/docs/range-sources/#thread-safety).

Reply:
Makes sense. I'll do that